### PR TITLE
WithNames format fixes

### LINKED ISSUE #12504: CSVWithNames via HTTP lacks header row on 20.4.7.67
**Describe the bug**
When using the HTTP interface, the CSVWithNames result format looks like just CSV, lacking the header row. It behaves properly via `clickhouse-client` though.

**How to reproduce**
```bash
~$ echo "select * from system.build_options
> where name='VERSION_DESCRIBE' format CSVWithNames" |
> curl localhost:8123 --data-binary @-
"VERSION_DESCRIBE","v20.4.7.67-stable"
```
```sql
host :) select * from system.build_options where name='VERSION_DESCRIBE' format CSVWithNames

SELECT *
FROM system.build_options
WHERE name = 'VERSION_DESCRIBE'
FORMAT CSVWithNames

"name","value"
"VERSION_DESCRIBE","v20.4.7.67-stable"
```

**Expected behavior**
The two results should match, and both have the header row with field names.

**Additional context**
TSVWithNames seems to work OK on both interfaces. Both were working on the last 20.4.6... version.

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix missing or excessive headers in `TSV/CSVWithNames` formats. This fixes #12504

- Fix `File(TSVWithNames*)` (header was written multiple times)
- Fix `clickhouse-local --format CSVWithNames*` (lack of header, broken after #12197)
- Fix `clickhouse-local --format CSVWithNames*` with zero rows (lack of header)

Continuation of: #12197
Fixes: #12504

<details>

HEADs:
- 8db28001e3d903bf824788ae989d888dd8db5bcb
- f58e4bf96a56aeffa2067d114b506fa9b3ef7c72

</details>

### CODE REVIEW

File: tests/queries/0_stateless/01375_output_format_with_names.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+function produce()
+{
+    echo 1
+    echo 1
+}
+
+echo 'TSVWithNames'
+produce | ${CLICKHOUSE_LOCAL} "${opts[@]}" --format TSVWithNames
```
Comment (akuzm) [inline tests/queries/0_stateless/01375_output_format_with_names.sh:19]: Is the `produce` part important, or `select '1' key from numbers(2) format TSVWithNames ` would also work?

**Review by akuzm — APPROVED** [review]

**Comment by akuzm** [comment]
@nikitamikhaylov FYI
You'll probably have to change all this when adding parallel formatting, so it's good that we have more test cases now.

File: tests/queries/0_stateless/01375_output_format_with_names.sh (hunk 1 above)
Comment (azat) [reply tests/queries/0_stateless/01375_output_format_with_names.sh:19]: egh, was too enthusiastic about scripting, will use your suggestion

**Comment by nikitamikhaylov** [comment]
> @nikitamikhaylov FYI
> 
> You'll probably have to change all this when adding parallel formatting, so it's good that we have more test cases now.

Thanks! I will think.

**Comment by akuzm** [comment]
I'll take a look at the perf test -- probably it depends on the old broken behavior and doesn't expect the header somewhere.

**Comment by akuzm** [comment]
The reason the performance test fails is that now we can't read an empty file as TSVWithNamesAndTypes:

```
$ clickhouse-local --query "
   create table t engine File(TSVWithNamesAndTypes, '1.tsv') as
      select * from numbers(1) where number < 0;
   select * from t;
"

Code: 117, e.displayText() = DB::Exception: Unknown field found in TSV header: '' at position 0
Set the 'input_format_skip_unknown_fields' parameter explicitly to ignore and proceed: Error while reading from '1.tsv': While executing File (version 20.7.1.1)
```

Can we add some support for this?

**Comment by azat** [comment]
>The reason the performance test fails is that now we can't read an empty file as TSVWithNamesAndTypes:

Not really, the problem is that the header wasn't written, while it should.

P.S. and reading empty files is not supported anyway, this is the result w/o this PR:

```sql
SELECT * FROM file('/dev/null', TSVWithNamesAndTypes, 'key Int')
Received exception from server (version 20.6.1):
Code: 117. DB::Exception: Received from localhost:9000. DB::Exception: Unknown field found in TSV header: '' at position 0
Set the 'input_format_skip_unknown_fields' parameter explicitly to ignore and proceed: While executing File.
```

**Comment by azat** [comment]
>the problem is that the header wasn't written, while it should.

@akuzm fixed

**Comment by artbeglaryan** [comment]
@akuzm will this be backported to 20.3 too? On original task https://github.com/ClickHouse/ClickHouse/pull/12197 it marked as 20.3-backported

```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.3.17.173 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.3.17 revision 54433.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
1
```

Also didn't work on latest 20.4 too

20.4
```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.4.9.110 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.4.9 revision 54434.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
1
```

20.5 is ok
```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.5.5.74 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.5.5 revision 54435.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
"T"
1
```

**Pull request merged**